File: lazyseq.py
def bounce(val):
    while callable(val):
        val = val()
    return val

def seq(source):
    return source if isinstance(source, Seq) else Seq.from_array(list(source))


class Seq:
    def __init__(self, *args):
        if len(args) == 2 and callable(args[1]):
            self.first  = args[0]
            self.__rest = args[1]
        else:
            if len(args) == 0:
                raise ArgumentError("Use None for empty sequences.")

            if callable(args[-1]):
                if len(args) == 1:
                    s = args[0]()
                else:
                    s = seq(args[:-2]).concat(Seq(*args[-2:]))
            else:
                s = seq(args)
                
            self.first = s.first
            self.__rest = lambda : s.rest

    @property
    def rest(self):
        self.rest = self.__rest and self.__rest()
        return self.rest

    def forced(self):
        step = lambda s: s and (lambda : step(s.rest))
        bounce(step(self.rest))
        return self

    def __iter__(self):
        s = self
        while s:
            yield s.first
            s = s.rest

    @classmethod
    def from_array(cls, a, i = 0):
        if i < len(a):
            return Seq(a[i], lambda : cls.from_array(a, i + 1))

    @classmethod
    def up_from(cls, start):
        return Seq(start, lambda : cls.up_from(start + 1))

    @classmethod
    def down_from(cls, start):
        return Seq(start, lambda : cls.down_from(start - 1))

    @classmethod
    def range(cls, start, limit):
        if limit >= start:
            return Seq.up_from(start).take_while(lambda x: x <= limit)
        else:
            return Seq.down_from(start).take_while(lambda x: x >= limit)

    @classmethod
    def constant(cls, val):
        return Seq(val, lambda : cls.constant(val))

    @classmethod
    def iterate(cls, x, f):
        return Seq(x, lambda : cls.iterate(f(x), f))
    
    def __str__(self):
        return " -> ".join(self.map(str))

    def size(self):
        def step(s, n):
            return (lambda : step(s.rest, n + 1)) if s else n
        return bounce(step(self, 0))

    def last(self):
        def step(s):
            return (lambda : step(s.rest)) if s.rest else s.first
        return bounce(step(self))

    def reverse(self):
        def step(rev, s):
            if s:
                return lambda : step(Seq(s.first, lambda : rev), s.rest)
            else:
                return rev
        return bounce(step(None, self))

    def take(self, n):
        if n > 0:
            return Seq(self.first,
                       lambda : self.rest and self.rest.take(n-1))

    def take_while(self, pred):
        if pred(self.first):
            return Seq(self.first,
                       lambda : self.rest and self.rest.take_while(pred))

    def drop(self, n):
        def step(s, n):
            if s and n > 0:
                return lambda : step(s.rest, n-1)
            else:
                return s
        return bounce(step(self, n))

    def drop_until(self, pred):
        def step(s):
            if s and not pred(s.first):
                return lambda : step(s.rest)
            else:
                return s
        return bounce(step(self))

    def pick(self, n):
        return self.drop(n).first

    def cycle(self):
        return self.cycle_from(self)

    def cycle_from(self, s):
        if s:
            return Seq(s.first, lambda : self.cycle_from(s.rest))
        else:
            return self.cycle()

    def select(self, pred):
        if pred(self.first):
            return Seq(self.first, lambda : self.rest and self.rest.select(pred))
        elif self.rest:
            r = self.rest.drop_until(pred)
            return r and r.select(pred)

    def distinct(self, back = None):
        if back and back.contains(self.first):
            r = self.drop_until(lambda x: not back.contains(x))
            return r and r.distinct(back)
        else:
            return Seq(self.first,
                       lambda : self.rest and self.rest.distinct(
                           Seq(self.first, lambda : back)))

    def find(self, pred):
        good = self.drop_until(pred)
        return good and good.first

    def contains(self, val):
        return self.drop_until(lambda x: x == val) is not None

    def forall(self, pred):
        return not self.drop_until(lambda x: not pred(x))

    def map(self, f):
        return Seq(f(self.first), lambda : self.rest and self.rest.map(f))

    def reduce(self, start, op):
        def step(val, s):
            if s:
                return lambda : step(op(val, s.first), s.rest)
            else:
                return val
        return bounce(step(start, self))

    def sum(self):
        return self.reduce(0, lambda a, b: a + b)

    def product(self):
        return self.reduce(1, lambda a, b: a * b)

    def fold(self, op):
        if self.rest:
            return self.rest.reduce(self.first, op)
        else:
            return self.first

    def min(self):
        return self.fold(lambda a, b: b if b < a else a)

    def max(self):
        return self.fold(lambda a, b: b if b > a else a)

    def zip_seq(self):
        firsts = self.map(lambda s: s and s.first)
        if not firsts.forall(lambda s: s is None):
            return Seq(firsts,
                       lambda : self.map(lambda s: s and s.rest).zip_seq())

    def sequentialize_with(self, *args):
        return Seq(self, lambda: args and seq(args).map(seq))

    def zip(self, *others):
        return self.sequentialize_with(*others).zip_seq()

    def combine(self, op, *others):
        return self.zip(*others).map(lambda s: s and s.fold(op))

    def add(self, *others):
        return self.combine((lambda a, b: a + b), *others)

    def __add__(self, other):
        return self.add(other)

    def sub(self, *others):
        return self.combine((lambda a, b: a - b), *others)

    def __sub__(self, other):
        return self.sub(other)

    def mul(self, *others):
        return self.combine((lambda a, b: a * b), *others)

    def __mul__(self, other):
        return self.mul(other)

    def div(self, *others):
        return self.combine((lambda a, b: a / b), *others)

    def __div__(self, other):
        return self.div(other)

    def eq(self, *others):
        return self.zip(*others).forall(
            lambda s: s.forall(lambda x: x == s.first) if s else true)

    def __eq__(self, other):
        try:
            return self.eq(other)
        except TypeError:
            False

    def __ne__(self, other):
        return not self.eq(other)

    def lazy_concat(self, s):
        return Seq(self.first,
                   lambda : self.rest.lazy_concat(s) if self.rest else s())

    def flatten(self):
        if self.rest:
            return self.first.lazy_concat(lambda : self.rest.flatten())
        else:
            return self.first

    def flat_map(self, fun):
        return self.map(fun).flatten()

    def concat(self, *others):
        return self.sequentialize_with(*others).flatten()

    def interleave_seq(self):
        head = lambda s: s.map(lambda t: t.first)
        tail = lambda s: s.map(lambda t: t.rest).interleave_seq()
        
        alive = self.select(lambda s: s is not None)
        if alive:
            return head(alive).lazy_concat(lambda : tail(alive))

    def interleave(self, *others):
        return self.sequentialize_with(*others).interleave_seq()

    def cartesian_seq(self):
        if self.rest:
            return self.first.flat_map(
                lambda s: self.rest.cartesian_seq().map(
                    lambda t: Seq(s, lambda : t)))
        else:
            return self.first.map(lambda s: Seq(s))

    def cartesian(self, *others):
        return self.sequentialize_with(*others).cartesian_seq()

    def cantor_fold(self, back, remaining):
        if remaining:
            t = Seq(remaining.first, lambda : back)
            z = self.zip(t).take_while(lambda x: x and x.pick(1)).flat_map(
                lambda x: x.pick(1).map(lambda y: Seq(x.first, lambda : y)))
            return Seq(z, lambda : self.cantor_fold(t, remaining.rest))

    def cantor_runs(self):
        if self.rest:
            return self.first.cantor_fold(None, self.rest.cantor_runs())
        else:
            return self.first.map(lambda x: Seq(Seq(x)))

    def cantor(self, *others):
        return self.sequentialize_with(*others).cantor_runs().flatten()

    def subseqs(self):
        return Seq(self, lambda : self.rest and self.rest.subseqs())

    def consec(self, n):
        return self.subseqs().map(lambda s: list(s.take(n)))


if __name__ == "__main__":
    s = seq("the quick brown fox jumps over".split())
    print "Sequence:     ", s
    print "Forced:       ", s.forced()
    print "Mangle last:  ",
    print s.subseqs().map(
        lambda sub: sub.first if sub.rest else sub.first.upper())
    print "Size:         ", s.size()
    print "Last:         ", s.last()
    print "Runs of 3:    ", s.consec(3).drop(3)
    print "Letter counts:", dict(s.map(lambda w: (w, len(w))).take(4))
    print "Repeat third: ", Seq.constant(s.pick(2)).take(5)
    print "Cycle:        ", s.cycle().take(8)
    print "Start at fox: ", s.drop_until(lambda x: x == "fox")
    print "Five letters: ", s.select(lambda x: len(x) == 5)
    print "First with r: ", s.find(lambda x: 'r' in x)
    print "All 3 letters:", s.forall(lambda x: len(x) == 3)
    print "Reverse:      ", s.reverse()
    print "Min and max:  ", "%s, %s" % (s.min(), s.max())
    print "Zip with ints:", s.zip('abcdefg').map(list).drop(3)
    print
    print "Number range: ", s.range(10, 20)
    print "Its sum:      ", s.range(10, 20).sum()
    print "Its product:  ", s.range(10, 20).product()
    print "flat_map:     ", Seq.range(4, 1).flat_map(lambda n: Seq.range(1, n))
    print "Iterate:      ", Seq.iterate(1, lambda x: 2 * x).take(10)
    print
    fib = Seq(0, 1, lambda : fib.rest + fib)
    print "Fibonacci:    ", fib.take(12)
    print "Compare:      ", (fib.take(10) == (0, 1, 1, 2, 3, 5, 8, 13, 21, 34))
    print "Compare:      ", (fib.take(10) == (0, 1, 1, 2, 3, 5, 8.2, 13, 21, 34))
    print
    print "No first:     ", Seq(lambda : s)
    print "One first:    ", Seq(1, lambda : s)
    print "Two firsts:   ", Seq(1, 2, lambda : s)
    print "Three firsts: ", Seq(1, 2, 3, lambda : s)
    print
    primes = Seq.up_from(2).select(
        lambda n: n < 4 or (primes.take_while(lambda m: m * m <= n).
                            forall(lambda m: n % m)))
    print "Prime numbers:", primes.take(10)
    print
    print "Concatenation:", s.take(3).concat(fib.take(2), primes.take(3))
    print "Interleave:   ", s.take(3).interleave(fib.take(2), primes.take(3))
    print "Cartesian:    ", fib.take(2).cartesian(primes.take(2), [0]).map(list)
    print "Cantor:       ", primes.cantor(primes, primes).take(5).map(list)
    print "Distinct:     ", fib.interleave(primes).distinct().take(10)

File: lazyseq.rb
def bounce(val)
  while val.is_a? Proc
    val = val.call
  end
  val
end


class Array
  def to_seq
    Seq.from_array self
  end
end


module Enumerable
  def to_seq
    to_a.to_seq
  end
end


class Seq
  attr_reader :first

  def initialize(*elems, &rest)
    case elems.length
    when 0
      seq = rest.call()
      @first = seq.first
      @rest  = lambda { seq.rest }
    when 1
      @first = elems[0]
      @rest  = rest
    when 2
      @first = elems[0]
      @rest  = lambda { Seq.new elems[1], &rest }
    else
      @first = elems[0]
      @rest  = lambda { elems[1..-2].to_seq.concat Seq.new(elems[-1], &rest) }
    end
  end

  def rest
    @rest = @rest && @rest.call()
    class << self; def rest; @rest end end
    freeze
    @rest
  end

  def forced
    step = lambda { |seq| lambda { step.call seq.rest } if seq }
    bounce step.call(self.rest)
    self
  end

  def to_enum
    Enumerator.new do |yielder|
      s = self
      while s
        yielder << s.first
        s = s.rest
      end
    end
  end

  def to_seq
    self
  end

  def to_a
    to_enum.to_a
  end

  def to_hash
    hash = {}
    to_enum.each { |k, v| hash[k] = v }
    hash
  end

  def self.from_array(array, i = 0)
    Seq.new(array[i]) { from_array array, i+1 } if i < array.length
  end

  def self.up_from(start)
    Seq.new(start) { up_from start.next }
  end

  def self.down_from(start)
    Seq.new(start) { down_from start.pred }
  end

  def self.range(start, limit)
    if limit >= start
      Seq.up_from(start).take_while { |x| x <= limit }
    else
      Seq.down_from(start).take_while { |x| x >= limit }
    end
  end

  def self.constant(val)
    Seq.new(val) { constant val }
  end

  def self.iterate(x, &f)
    Seq.new(x) { iterate f.call(x), &f }
  end

  def to_s
    map(&:to_s).to_a.join ' -> '
  end

  def size
    step = lambda { |seq, n|
      if seq then lambda { step.call(seq.rest, n + 1) }  else n end
    }
    bounce step.call(self, 0)
  end

  def last
    step = lambda { |seq|
      if seq.rest then lambda { step.call seq.rest } else seq.first end
    }
    bounce step.call(self)
  end

  def reverse
    step = lambda { |rev, seq|
      if seq
        lambda { step.call Seq.new(seq.first) { rev }, seq.rest }
      else
        rev
      end
    }
    bounce step.call(nil, self)
  end

  def take(n)
    Seq.new(self.first) { rest.take n-1 if rest } if n > 0
  end

  def take_while(&pred)
    Seq.new(first) { rest.take_while &pred if rest } if pred.call(first)
  end

  def drop(n)
    step = lambda { |seq, n|
      if seq and n > 0 then lambda { step.call seq.rest, n-1 } else seq end
    }
    bounce step.call(self, n)
  end

  def drop_until(&pred)
    step = lambda { |seq|
      if seq and not pred.call seq.first then
        lambda { step.call seq.rest }
      else
        seq
      end
    }
    bounce step.call(self)
  end

  def pick(n)
    drop(n).first
  end

  def cycle
    cycle_from self
  end

  def cycle_from(seq)
    if seq then Seq.new(seq.first) { cycle_from(seq.rest) } else cycle end
  end

  def select(&pred)
    if pred.call first
      Seq.new(first) { rest.select &pred if rest }
    elsif rest
      r = rest.drop_until &pred
      r.select &pred if r
    end
  end

  def distinct(back = nil)
    if back and back.contains? first
      r = drop_until { |x| not back.contains? x }
      r.distinct(back) if r
    else
      Seq.new(first) { rest.distinct Seq.new(first) { back } if rest }
    end
  end

  def find(&pred)
    if good = drop_until(&pred) then good.first end
  end

  def contains?(val)
    not drop_until { |x| x == val }.nil?
  end

  def forall?(&pred)
    not drop_until { |x| not pred.call x }
  end

  def map(&f)
    Seq.new(f.call self.first) { rest.map(&f) if rest }
  end

  def reduce(start = nil, &op)
    step = lambda do |val, seq|
      if seq
        lambda { step.call op.call(val, seq.first), seq.rest }
      else
        val
      end
    end
    bounce step.call(start, self)
  end

  def sum
    reduce(0) { |a, b| a + b }
  end

  def product
    reduce(1) { |a, b| a * b }
  end

  def fold(&op)
    if rest then rest.reduce first, &op else first end
  end

  def min
    fold { |a, b| b < a ? b : a }
  end

  def max
    fold { |a, b| b > a ? b : a }
  end

  def zip_seq
    firsts = map { |s| s.first if s }
    unless firsts.forall? &:nil?
      Seq.new(firsts) { map { |s| s.rest if s }.zip_seq }
    end
  end

  def sequentialize_with(*args)
    Seq.new(self) { args.to_seq.map { |s| s.to_seq } if args.length > 0 }
  end

  def zip(*others)
    sequentialize_with(*others).zip_seq
  end

  def combine(*others, &op)
    zip(*others).map { |seq| seq.fold &op if seq }
  end

  def +(*others)
    combine(*others) { |a, b| a + b }
  end

  def -(*others)
    combine(*others) { |a, b| a - b }
  end

  def *(*others)
    combine(*others) { |a, b| a * b }
  end

  def /(*others)
    combine(*others) { |a, b| a / b }
  end

  def ==(*others)
    zip(*others).forall? do |seq|
      if seq then seq.forall? { |x| x == seq.first } else true end
    end
  end

  def lazy_concat(&seq)
    Seq.new(first) { if rest then rest.lazy_concat(&seq) else seq.call end }
  end

  def flatten
    if rest then first.lazy_concat { rest.flatten } else first end
  end

  def flat_map(&fun)
    map(&fun).flatten
  end

  def concat(*others)
    sequentialize_with(*others).flatten
  end

  def interleave_seq
    alive = select { |seq| not seq.nil? }
    alive.map(&:first).lazy_concat { alive.map(&:rest).interleave_seq } if alive
  end

  def interleave(*others)
    sequentialize_with(*others).interleave_seq
  end

  def cartesian_seq
    if rest
      first.flat_map { |s| rest.cartesian_seq.map { |t| Seq.new(s) { t } } }
    else
      first.map { |s| Seq.new s }
    end
  end

  def cartesian(*others)
    sequentialize_with(*others).cartesian_seq
  end

  def cantor_fold(back, remaining)
    if remaining
      t = Seq.new(remaining.first) { back }
      z = zip(t).take_while { |x| x and x.pick 1 }.flat_map do |x|
        a = x.first
        x.pick(1).map { |y| Seq.new(a) { y } }
      end
      Seq.new(z) { cantor_fold t, remaining.rest }
    end
  end

  def cantor_runs
    if rest
      first.cantor_fold nil, rest.cantor_runs
    else
      first.map { |x| Seq.new Seq.new x }
    end
  end

  def cantor(*others)
    sequentialize_with(*others).cantor_runs.flatten
  end

  def subseqs
    Seq.new(self) { rest.subseqs if rest }
  end

  def consec(n)
    subseqs.map { |s| s.take(n).to_a }
  end

  def self.tree_walk(root, next_level)
    next_step = lambda { |path|
      if path
        top = path.first
        s = next_level.call top.first
        if s
          Seq.new(s) { path }
        elsif top.rest
          Seq.new(top.rest) { path.rest }
        elsif path.rest
          backtrack = path.rest.drop_until { |s| s.rest }
          Seq.new(backtrack.first.rest) { backtrack.rest } if backtrack
        end
      end
    }

    Seq.iterate(Seq.new(Seq.new(root)), &next_step).take_while { |x|
      x }.map { |x| x.first.first }
  end
end


if __FILE__ == $0
  seq = %w{the quick brown fox jumps over}.to_seq
  puts "Sequence:      #{seq}"
  puts "Forced:        #{seq.forced}"
  puts "Mangle last:   " +
    seq.subseqs.map { |sub| sub.rest ? sub.first : sub.first.upcase }.to_s
  puts "Size:          #{seq.size}"
  puts "Last:          #{seq.last}"
  puts "Runs of 3:     #{seq.consec(3).drop(3)}"
  puts "Letter counts: #{seq.map { |w| [w, w.length] }.take(4).to_hash}"
  puts "Repeat third:  #{Seq.constant(seq.pick(2)).take 5}"
  puts "Cycle:         #{seq.cycle.take 8}"
  puts "Start at fox:  #{seq.drop_until { |x| x == 'fox' }}"
  puts "Five letters:  #{seq.select { |x| x.length == 5 }}"
  puts "First with r:  #{seq.find { |x| x.include? 'r' }}"
  puts "All 3 letters: #{seq.forall? { |x| x.length == 3 }}"
  puts "Reverse:       #{seq.reverse}"
  puts "Min and max:   #{seq.min}, #{seq.max}"
  puts "With indexes:  #{seq.zip('abcdefg'.chars).map(&:to_a).drop 3}"
  puts
  puts "Number range:  #{Seq.range 10, 20}"
  puts "Its sum:       #{Seq.range(10, 20).sum}"
  puts "Its product:   #{Seq.range(10, 20).product}"
  puts "flat_map:      #{Seq.range(4, 1).flat_map { |n| Seq.range(1, n) }}" 
  puts "String range:  #{Seq.range "ady", "aeg"}"
  puts "Iterate:       #{Seq.iterate(1) { |x| 2 * x }.take(10)}"
  puts
  fib = Seq.new(0, 1) { fib.rest + fib }
  puts "Fibonacci:     #{fib.take 12}"
  puts "Compare:       #{fib.take(10) == [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]}"
  puts "Compare:       #{fib.take(10) == [0, 1, 1, 2, 3, 5, 8.2, 13, 21, 34]}"
  puts
  puts "No first:      #{Seq.new() { seq }}"
  puts "One first:     #{Seq.new(1) { seq }}"
  puts "Two firsts:    #{Seq.new(1, 2) { seq }}"
  puts "Three firsts:  #{Seq.new(1, 2, 3) { seq }}"
  puts
  primes = Seq.up_from(2).select do |n|
    n < 4 or primes.take_while { |m| m * m <= n }.forall? { |m| n % m > 0 }
  end
  puts "Prime numbers: #{primes.take(10)}"
  puts
  puts "Concatenation: #{seq.take(3).concat(fib.take(2), primes.take(3))}"
  puts "Interleave:    #{seq.take(3).interleave(fib.take(2), primes.take(3))}"
  puts "Cartesian:     #{fib.take(2).cartesian(primes.take(2), [0]).map &:to_a}"
  puts "Cantor:        #{primes.cantor(primes, primes).take(5).map &:to_a}"
  puts "Distinct:      #{fib.interleave(primes).distinct.take(10)}"
  puts

  def permutations(degree)
    next_level = lambda { |perm|
      i = perm.index nil
      unless i.nil?
        Seq.range(1, degree).select { |n| not perm.include? n }.map { |to|
          perm[0..i-1] + [to] + perm[i+1..-1] }
      end
    }

    Seq.tree_walk([0] + Array.new(degree), next_level).map { |p|
      p[1..-1] }.select { |p| not p.include? nil }
  end

  puts "Permutations:  #{permutations 4}"
end
